x/supplier/keeper/msg_server_stake_supplier.go

```diff
@@ -140,15 +124,46 @@ func (k msgServer) StakeSupplier(ctx context.Context, msg *suppliertypes.MsgStak
 		return nil, status.Error(codes.InvalidArgument, err.Error())
 	}
 
-	// Send the coins from the message signer account to the staked supplier pool
 	supplierStakingFee := k.GetParams(ctx).StakingFee
-	stakeWithFee := sdk.NewCoins(coinsToEscrow.Add(*supplierStakingFee))
-	err = k.bankKeeper.SendCoinsFromAccountToModule(ctx, msgSignerAddress, suppliertypes.ModuleName, stakeWithFee)
-	if err != nil {
-		logger.Info(fmt.Sprintf("ERROR: could not send %v coins from %q to %q module account due to %v", coinsToEscrow, msgSignerAddress, suppliertypes.ModuleName, err))
-		return nil, status.Error(codes.InvalidArgument, err.Error())
+
+	// The Supplier is increasing its stake, so escrow the difference
+	if supplierCurrentStake.Amount.LT(msg.Stake.Amount) {
+		coinsToEscrow := msg.Stake.Sub(supplierCurrentStake)
+		stakeWithFee := sdk.NewCoins(coinsToEscrow.Add(*supplierStakingFee))
+
+		// Send the coins from the message signer account to the staked supplier pool
+		err = k.bankKeeper.SendCoinsFromAccountToModule(ctx, msgSignerAddress, suppliertypes.ModuleName, stakeWithFee)
+		if err != nil {
+			logger.Info(fmt.Sprintf("ERROR: could not send %s coins from %q to %q module account due to %v", coinsToEscrow, msgSignerAddress, types.ModuleName, err))
+			return nil, status.Error(codes.InvalidArgument, err.Error())
+		}
+
+		logger.Info(fmt.Sprintf("Successfully escrowed %s from %q to %q module account", coinsToEscrow, msgSignerAddress, types.ModuleName))
+	}
+
+	// The supplier is decreasing its stake, unescrow the difference but deduct the staking fee.
+	if supplierCurrentStake.Amount.GT(msg.Stake.Amount) {
+		coinsToUnescrow := supplierCurrentStake.Sub(*msg.Stake)
+
+		if coinsToUnescrow.Amount.LTE(supplierStakingFee.Amount) {
+			err = types.ErrSupplierInvalidStake.Wrapf(
+				"supplier with owner %q must unescrow more than supplier staking fee (%s)",
+				msg.GetOwnerAddress(), supplierStakingFee,
+			)
+			logger.Info(fmt.Sprintf("ERROR: %s", err))
+			return nil, status.Error(codes.InvalidArgument, err.Error())
+		}
+
+		coinsAfterFee := sdk.NewCoins(coinsToUnescrow.Sub(*supplierStakingFee))
+
+		// Send the coins from the staked supplier pool to the message signer account
+		if err := k.bankKeeper.SendCoinsFromModuleToAccount(ctx, types.ModuleName, msgSignerAddress, coinsAfterFee); err != nil {
+			logger.Info(fmt.Sprintf("ERROR: could not send %s coins from %q module account to %q due to %v", coinsToUnescrow, types.ModuleName, msgSignerAddress, err))
+			return nil, status.Error(codes.InvalidArgument, err.Error())
+		}
+
+		logger.Info(fmt.Sprintf("Successfully unescrowed %s from %q module account to %q", coinsToUnescrow, types.ModuleName, msgSignerAddress))
```

Comment:
Where is the MinStake check happening to make sure there are no problems?

Reply:
At this point in the code the new stake amount is already checked against min stake.

The check is done here: https://github.com/pokt-network/poktroll/pull/946/files#diff-a21ef4c8a0a17759b5b114f280982371695f9467e42fbbc237bf29bbac43e784R111